objectivec/Tests/GPBMessageTests.m

```diff
@@ -431,6 +474,44 @@ - (void)testKVC_SetValue_ForKey {
   [self assertAllFieldsKVCMatch:message];
 }
 
+- (void)testKVOBasic {
+  TestAllTypes *message = [TestAllTypes message];
+  GPBKVOTestObserver *observer =
+      [[[GPBKVOTestObserver alloc] initWithObservee:message
+                                            keyPath:@"optionalString"]
+       autorelease];
+  XCTAssertFalse(observer.didObserve);
+  message.defaultString = @"Hello";
+  XCTAssertFalse(observer.didObserve);
+  message.optionalString = @"Hello";
+  XCTAssertTrue(observer.didObserve);
+}
+
+- (void)testKVOAutocreate {
+  TestAllTypes *message = [TestAllTypes message];
+  GPBKVOTestObserver *autocreateObserver =
+      [[[GPBKVOTestObserver alloc] initWithObservee:message
+                                            keyPath:@"optionalImportMessage"]
+       autorelease];
+ GPBKVOTestObserver *innerFieldObserver =
+     [[[GPBKVOTestObserver alloc] initWithObservee:message
+                                           keyPath:@"optionalImportMessage.d"]
+      autorelease];
+  XCTAssertFalse(autocreateObserver.didObserve);
+  XCTAssertFalse(innerFieldObserver.didObserve);
+  message.optionalImportMessage.d = 2;
+
+  // Autocreation of fields is not observed by KVO.
+  XCTAssertFalse(autocreateObserver.didObserve);
```

Comment:
So it sorta feels like this should have fired, after all the field got set.

You should likely also first just fetch the value of `d` (i.e. - get the default value), and ensure that doesn't trigger either of them.